gxutil/pm.go

```diff
@@ -775,3 +670,208 @@ func setInstallPathCache(env string, global bool, val string) {
+	const maxGoroutines = 20
+	// TODO: Consider making this value a parameter of the function
+	// (or an attribute of the `PM` structure).
+
+	// Central queue of dependencies that need to be fetched. Handled only
+	// by this function. Created with an initial a capacity on the rough
+	// estimate of twice the maximum goroutines running.
+	depQueue := NewDependencyQueue(maxGoroutines * 2)
+
+	// List of channels for each spawned goroutine to store either the
+	// fetched package or an error. To ensure they are non blocking the
+	// maximum number of goroutines it's assigned for their capacity
+	// (worst case scenario).
+	fetchedPackages := make(chan *Package, maxGoroutines)
+	fetchErrs := make(chan error, maxGoroutines)
+
+	// To start the process add the dependencies of the root package.
+	addedDepCount := depQueue.AddPackageDependencies(pkg)
+	pm.ProgMeter.AddTodos(addedDepCount)
+
+	// Counter to keep track of spawned goroutines, it's not locked as it's
+	// only handled by this function. Decremented any time a message is received
+	// on the above channels, which indicates that the goroutine has finished.
+	activeGoroutines := 0
+
+	// Main loop of the function.
+	for {
+		// If there are no more dependencies to install and there aren't any
+		// goroutines running (which could potentially add new dependencies
+		// to the queue) we're finished.
+		if depQueue.Len() == 0 && activeGoroutines == 0 {
+			return nil
+		}
+
+		// Keep spawning goroutines until the allowed maximum or until
+		// there are no new dependencies to fetch at the moment.
+		for activeGoroutines < maxGoroutines && depQueue.Len() > 0 {
+
+			// Goroutine that only calls `GetPackageTo` to fetch the dependency,
+			// it either returns it or returns an error.
+			go func(dep *Dependency) {
+
+				pkgDir := filepath.Join(location, "gx", "ipfs", dep.Hash)
+				// TODO: Encapsulate in a function. Used in too many places
+				// and is part of the standard.
+
+				pm.ProgMeter.AddEntry(dep.Hash, dep.Name, "[fetch]   <ELAPSED>"+dep.Hash)
+				pkg, err := pm.GetPackageTo(dep.Hash, pkgDir)
+
+				// Either with error or with the package the goroutine ends here.
+				if err != nil {
+					fetchErrs <- fmt.Errorf("failed to fetch package: %s: %s", dep.Hash, err)
+					pm.ProgMeter.Error(dep.Hash, err.Error())
+				} else {
+					fetchedPackages <- pkg
+					pm.ProgMeter.Finish(dep.Hash)
+				}
+			}(depQueue.GetDependency())
+
+			activeGoroutines++
+		}
+
+		// Once all the possible goroutines have been spawned wait
+		// for anyone to finish and analyze (restart loop) if more
+		// goroutines can be called.
+		select {
+		case fetchedPkg := <-fetchedPackages:
+			VLog("fetched dep: %s", fetchedPkg.Name)
+			addedDepCount := depQueue.AddPackageDependencies(fetchedPkg)
+			pm.ProgMeter.AddTodos(addedDepCount)
+			activeGoroutines--
+		case err := <-fetchErrs:
+			Error("parallel fetch: %s", err)
+			return err
+			// The rest of the goroutines already running are guaranteed to have
```

Comment:
We should still wait. Otherwise, we'll quit the program while they're still running (and we'll end up with partial packages).

Reply:
Good point, I'll fix that.

Reply:
I'm thinking here that if I have an error I break out of the main `for` and do a separate one just to receive from the channels until no more goroutines are running (but do nothing else, don't move the progression meter). In this case I would return the first error received from the goroutines and the others would be discarded (just reported through the `Error()` log but not returned to the user). Is that ok?

Reply:
SGTM.